File: macos/Sources/ElevateCore/Support/ErrorLog.swift
import Foundation

/// A single recorded error, with its timestamp.
public struct DiagnosticsError: Sendable, Equatable {
    public let date: Date
    public let message: String

    public init(date: Date, message: String) {
        self.date = date
        self.message = message
    }
}

/// A ring buffer of the most recent errors, capped at a fixed capacity.
public struct ErrorLog: Sendable, Equatable {
    private var buffer: [DiagnosticsError] = []
    private let capacity: Int

    public init(capacity: Int = 50) {
        self.capacity = capacity
    }

    /// Appends an error, evicting the oldest entry once over capacity.
    public mutating func append(_ message: String, at date: Date = .now) {
        buffer.append(DiagnosticsError(date: date, message: message))
        if buffer.count > capacity {
            buffer.removeFirst(buffer.count - capacity)
        }
    }

    /// Entries ordered oldest to newest.
    public var entries: [DiagnosticsError] {
        buffer
    }
}

File: windows/src/Elevate.Core/Support/ErrorLog.cs
namespace Elevate.Core.Support;

/// <summary>A single recorded error, with its timestamp.</summary>
public sealed record DiagnosticsError(DateTimeOffset Date, string Message);

/// <summary>A ring buffer of the most recent errors, capped at a fixed capacity. Port of the Swift <c>ErrorLog</c>.</summary>
public sealed class ErrorLog
{
    private readonly List<DiagnosticsError> _buffer = [];
    private readonly int _capacity;

    public ErrorLog(int capacity = 50)
    {
        ArgumentOutOfRangeException.ThrowIfLessThan(capacity, 1);
        _capacity = capacity;
    }

    /// <summary>Appends an error, evicting the oldest entry once over capacity.</summary>
    public void Append(string message, DateTimeOffset? at = null)
    {
        _buffer.Add(new DiagnosticsError(at ?? DateTimeOffset.UtcNow, message));
        if (_buffer.Count > _capacity)
        {
            _buffer.RemoveRange(0, _buffer.Count - _capacity);
        }
    }

    /// <summary>Entries ordered oldest to newest.</summary>
    public IReadOnlyList<DiagnosticsError> Entries => _buffer;

    /// <summary>Swift's <c>Equatable</c>: same capacity and the same entries in the same order.</summary>
    public bool ContentEquals(ErrorLog other)
    {
        ArgumentNullException.ThrowIfNull(other);
        return _capacity == other._capacity && _buffer.SequenceEqual(other._buffer);
    }
}
